logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/debug/RepairFaultRemovalKeysUseCase.kt

```diff
@@ -0,0 +1,163 @@
+import kotlinx.coroutines.withContext
+
+/**
+ * Use case to recover from an invalid removal key state on all MLS conversations.
+ * This is an experimental feature to be trigger manually via debug options for now.
+ */
+interface RepairFaultRemovalKeysUseCase {
+    suspend operator fun invoke(param: TargetedRepairParam): RepairResult
+}
+
+internal class RepairFaultRemovalKeysUseCaseImpl(
+    private val selfUserId: UserId,
+    private val conversationRepository: ConversationRepository,
+    private val resetMLSConversation: ResetMLSConversationUseCase,
+    private val transactionProvider: CryptoTransactionProvider,
+    private val dispatcher: KaliumDispatcher = KaliumDispatcherImpl,
+) : RepairFaultRemovalKeysUseCase {
+
+    val logger by lazy { kaliumLogger.withTextTag("RepairFaultRemovalKeysUseCase") }
+
+    override suspend fun invoke(param: TargetedRepairParam): RepairResult = withContext(dispatcher.io) {
+        if (selfUserId.domain != param.domain) {
+            logger.w(
+                "Attempted to repair faulty removal keys in domain ${param.domain}, but user belongs to ${selfUserId.domain}. Aborting."
+            )
+            return@withContext RepairResult.RepairNotNeeded
+        }
+
+        conversationRepository.getMLSConversationsByDomain(param.domain)
+            .flatMap { conversations ->
+                val conversationsWithFaultyKeys = conversations.filter { convo ->
+                    checkConversationHasFaultyKey(convo.protocol, param.faultyKey)
+                }
+
+                when {
+                    conversationsWithFaultyKeys.isEmpty() -> RepairResult.NoConversationsToRepair.right()
+                    else -> repairConversations(conversationsWithFaultyKeys, conversations.size).right()
+                }
+            }
+            .fold(
+                fnL = {
+                    logger.e("Error occurred during repair of faulty removal keys")
+                    RepairResult.Error
+                },
+                fnR = { it }
+            )
+    }
+
+    private suspend fun repairConversations(
+        conversationsToRepair: List<Conversation>,
+        totalChecked: Int
+    ): RepairResult {
+        val (successful, failed) = conversationsToRepair
+            .map { convo -> convo to delegateResetMLSConversation(convo.id) }
+            .partition { (_, result) -> result.isRight() }
+
+        return RepairResult.RepairPerformed(
+            totalConversationsChecked = totalChecked,
+            conversationsWithFaultyKeys = conversationsToRepair.size,
+            successfullyRepairedConversations = successful.size,
+            failedRepairs = failed.map { (convo, _) -> convo.id.toLogString() }
+        )
+    }
+
+    @Suppress("TooGenericExceptionCaught")
+    private suspend fun delegateResetMLSConversation(conversationId: ConversationId): Either<CoreFailure, Unit> = try {
+        transactionProvider.transaction("RepairFaultRemovalKeys") { context ->
```

Comment:
Instead of opening transaction to each conversation, open it in 
repairConversations and pass context here